### PR TITLE
Add auto-generated rules table to the readme

### PR DESCRIPTION
**Pull request by bmish**
This change replaces the manually-maintained list of rules in readme.md with an autogenerated rules table containing columns of information:
* ✅ if the rule is in the `recommended` config
* 🔧 if the rule is autofixable
* The rule name with a link to the rule doc
* A description of the rule

Benefits of the table:
* Much improved readability now that information is separated into columns.
* The source of truth for the rule descriptions has been moved into the rules themselves using the standard eslint [meta.docs.description](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics) property. This makes it easier for humans and tooling to find the rule descriptions, including the [eslint-plugin/require-meta-docs-description](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-docs-description.md) lint rule which we can enable as a follow-up to enforce the presence and consistency of rule descriptions.
* A new script `yarn generate-rules-table` will automatically regenerate the rules table so no more need to manually maintain it.
* I have added a CI test to ensure the rules table is always correct and up-to-date.
* This rules table design and script is based on the same design used by [eslint-plugin-qunit](https://github.com/platinumazure/eslint-plugin-qunit#rules), [eslint-plugin-ember](https://github.com/ember-cli/eslint-plugin-ember#-rules), and [ember-template-lint](https://github.com/ember-template-lint/ember-template-lint#rules).

### CODE REVIEW

File: scripts/update-rules.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// Config/preset/fixable emojis.
+const EMOJI_RECOMMENDED = ':white_check_mark:';
+const EMOJI_FIXABLE = ':wrench:';
+
+// Generate rule table contents.
+const rulesTableContent = Object.keys(rules).filter(ruleName => !rules[ruleName].meta.deprecated)
+	.sort()
+	.map(ruleName => {
+		// Check which emojis to show for this rule.
+		const isRecommended = Object.prototype.hasOwnProperty.call(configs.recommended.rules, `unicorn/${ruleName}`);
```
Comment (sindresorhus) [inline scripts/update-rules.js:22]: This check is faulty. It should not mark the rule as recommended if it's in the recommended preset as off.

File: scripts/update-rules.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const path = require('path');
+const {rules, configs} = require('../');
+
+const pathReadme = path.resolve(__dirname, '../README.md');
+const readmeContents = fs.readFileSync(pathReadme, 'utf8');
+const tablePlaceholder = /<!--RULES_TABLE_START-->[\S\s]*<!--RULES_TABLE_END-->/;
+
+// Config/preset/fixable emojis.
+const EMOJI_RECOMMENDED = ':white_check_mark:';
+const EMOJI_FIXABLE = ':wrench:';
```
Comment (sindresorhus) [inline scripts/update-rules.js:15]: Just use the actual emojis instead.

File: scripts/update-rules.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			isRecommended ? EMOJI_RECOMMENDED : '',
+			isFixable ? EMOJI_FIXABLE : ''
+		].join('');
+
+		const url = `docs/rules/${ruleName}.md`;
+		const link = `[${ruleName}](${url})`;
+
+		const {description} = rules[ruleName].meta.docs;
+
+		return `| ${emojis} | ${link} | ${description} |`;
```
Comment (sindresorhus) [inline scripts/update-rules.js:35]: I prefer the column ordering of https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#supported-rules

Also separate column for fixable.

File: scripts/update-rules.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+// Automatically regenerates the rules table in readme.md.
+
+const fs = require('fs');
+const path = require('path');
+const {rules, configs} = require('../');
+
+const pathReadme = path.resolve(__dirname, '../README.md');
```
Comment (sindresorhus) [inline scripts/update-rules.js:9]:
```suggestion
const pathReadme = path.resolve(__dirname, '../readme.md');
```

File: package.json [hunk 5]
```diff
@@ -18,7 +18,8 @@
 		"create-rule": "node ./scripts/create-rule.js",
 		"run-rules-on-codebase": "node ./test/run-rules-on-codebase/lint.mjs",
 		"integration": "node ./test/integration/test.js",
-		"smoke": "eslint-remote-tester --config ./test/smoke/eslint-remote-tester.config.js"
+		"smoke": "eslint-remote-tester --config ./test/smoke/eslint-remote-tester.config.js",
+		"update-rules": "node ./scripts/update-rules.js"
```
Comment (sindresorhus) [inline package.json:22]: `update-rules` is not exactly describing what it does.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by papb** [comment]
Interesting approach. There is another nice approach to this kind of automation that I plan to use soon in my projects: [eslint-plugin-codegen](https://www.npmjs.com/package/eslint-plugin-codegen).

File: scripts/update-rules.js (hunk 1 above)
Comment (bmish) [reply scripts/update-rules.js]: Good catch, fixed.

File: scripts/update-rules.js (hunk 2 above)
Comment (bmish) [reply scripts/update-rules.js]: Fixed.

File: scripts/update-rules.js (hunk 3 above)
Comment (bmish) [reply scripts/update-rules.js]: Updated as requested.

File: scripts/update-rules.js (hunk 4 above)
Comment (bmish) [reply scripts/update-rules.js]: Fixed.

File: package.json (hunk 5 above)
Comment (bmish) [reply package.json]: Renamed to `update-rules-in-readme`.

**Comment by bmish** [comment]
> Interesting approach. There is another nice approach to this kind of automation that I plan to use soon in my projects: [eslint-plugin-codegen](https://www.npmjs.com/package/eslint-plugin-codegen).

Thanks for bringing that up! Haven't heard of it before and it sounds promising. For now, I'd still prefer to get this change in with my simple Node script for generating the table, but would definitely consider swapping out that script for a better solution as a follow-up.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -47,4 +47,4 @@ jobs:
           node-version: 14
       - run: npm install
       - run: npm run integration
-      - run: npm run update-rules && git diff --exit-code
+      - run: npm run update-rules-in-readme && git diff --exit-code
```
Comment (fisker) [inline .github/workflows/main.yml:50]: How about `generate-rules-table`

Comment (bmish) [reply .github/workflows/main.yml]: Good idea, updated.

**Comment by fisker** [comment]
It will be great if you can also update [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/1ea39f4e5c8767093cbd7e52d9bea1f082c0ac64/scripts/create-rule.js#L140) too.

File: scripts/generate-rules-table.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+// Automatically regenerates the rules table in readme.md.
+
+const fs = require('fs');
+const path = require('path');
+const {rules, configs} = require('../');
+
+const pathReadme = path.resolve(__dirname, '../readme.md');
+const readmeContents = fs.readFileSync(pathReadme, 'utf8');
+const tablePlaceholder = /<!--RULES_TABLE_START-->[\S\s]*<!--RULES_TABLE_END-->/;
```
Comment (fisker) [inline scripts/generate-rules-table.js:11]: Will `dotAll`flag work?

File: scripts/generate-rules-table.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		return `| ${link} | ${description} | ${isRecommended ? EMOJI_RECOMMENDED : ''} | ${isFixable ? EMOJI_FIXABLE : ''} |`;
+	})
+	.join('\n');
+
+fs.writeFileSync(
+	pathReadme,
+	readmeContents.replace(
+		tablePlaceholder,
+		`<!--RULES_TABLE_START-->\n\n| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |\n|:--------|:--------|:---|:---|\n${rulesTableContent}\n\n<!--RULES_TABLE_END-->`
+	)
```
Comment (fisker) [inline scripts/generate-rules-table.js:39]: Let's use ESModule?

File: readme.md [hunk 9]
```diff
@@ -114,79 +114,90 @@ Configure it in `package.json`.
...
-- [prefer-type-error](docs/rules/prefer-type-error.md) - Enforce throwing `TypeError` in type checking conditions. *(fixable)*
-- [prevent-abbreviations](docs/rules/prevent-abbreviations.md) - Prevent abbreviations. *(partly fixable)*
-- [string-content](docs/rules/string-content.md) - Enforce better string content. *(fixable)*
-- [throw-new-error](docs/rules/throw-new-error.md) - Require `new` when throwing an error. *(fixable)*
+Each rule has emojis denoting:
+
+* ✅ if it belongs to the `recommended` configuration
+* 🔧 if some problems reported by the rule are automatically fixable by the `--fix` [command line](https://eslint.org/docs/user-guide/command-line-interface#fixing-problems) option
+
+<!--RULES_TABLE_START-->
```
Comment (fisker) [inline readme.md:122]: Let's add a comment before this, something like "It's auto-generated, don't modify"

**Review by fisker — CHANGES_REQUESTED** [review]

File: scripts/generate-rules-table.js (hunk 7 above)
Comment (bmish) [reply scripts/generate-rules-table.js:11]: What do you mean by that?

File: scripts/generate-rules-table.js (hunk 8 above)
Comment (bmish) [reply scripts/generate-rules-table.js:39]: You mean for the `writeFileSync` function? Updated.

File: readme.md (hunk 9 above)
Comment (bmish) [reply readme.md:123]: Good idea, added.

**Comment by bmish** [comment]
> It will be great if you can also update [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/1ea39f4e5c8767093cbd7e52d9bea1f082c0ac64/scripts/create-rule.js#L140) too.

@fisker I just updated the `create-rule` script to also update the rules table and fill in the rule description inside the rule.

File: scripts/generate-rules-table.js (hunk 8 above)
Comment (fisker) [reply scripts/generate-rules-table.js:39]: No, I mean rename this file to `generate-rules-table.mjs`

File: scripts/generate-rules-table.js (hunk 7 above)
Comment (fisker) [reply scripts/generate-rules-table.js:11]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/dotAll

File: scripts/generate-rules-table.js (hunk 8 above)
Comment (bmish) [reply scripts/generate-rules-table.js:39]: Updated to mjs.

File: scripts/generate-rules-table.js (hunk 7 above)
Comment (bmish) [reply scripts/generate-rules-table.js]: Updated to use `RegExp(..., 's')`.

File: scripts/generate-rules-table.mjs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+		return `| ${link} | ${description} | ${isRecommended ? EMOJI_RECOMMENDED : ''} | ${isFixable ? EMOJI_FIXABLE : ''} |`;
+	})
+	.join('\n');
+
+writeFileSync(
+	pathReadme,
+	readmeContents.replace(
+		tablePlaceholder,
+		`<!--RULES_TABLE_START-->\n\n| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |\n|:--------|:--------|:---|:---|\n${rulesTableContent}\n\n<!--RULES_TABLE_END-->`
```
Comment (fisker) [inline scripts/generate-rules-table.mjs:41]:
```suggestion
outdent`
<!--RULES_TABLE_START-->

| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |
|:--------|:--------|:---|:---|
${rulesTableContent}

<!--RULES_TABLE_END-->`
```

Comment (bmish) [reply scripts/generate-rules-table.mjs]: Good idea, updated.

File: scripts/generate-rules-table.mjs [hunk 11]
```diff
@@ -38,6 +39,13 @@ writeFileSync(
 	pathReadme,
 	readmeContents.replace(
 		tablePlaceholder,
-		`<!--RULES_TABLE_START-->\n\n| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |\n|:--------|:--------|:---|:---|\n${rulesTableContent}\n\n<!--RULES_TABLE_END-->`
+		outdent`
+		<!--RULES_TABLE_START-->
+
+		| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |
+		|:--------|:--------|:---|:---|
```
Comment (fisker) [inline scripts/generate-rules-table.mjs:46]: Let's use

```markdown
| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |
| :--- | :---------- | :------------------- | :--------------- |
```

OR

```markdown
| Name | Description | ${EMOJI_RECOMMENDED} | ${EMOJI_FIXABLE} |
| :-- | :-- | :-- | :-- |
```

Your choice

Comment (bmish) [reply scripts/generate-rules-table.mjs]: Updated to the first one.

File: test/package.mjs [hunk 12]
```diff
@@ -70,7 +70,7 @@ test('Every rule is defined in readme.md usage and list of rules in alphabetical
 	const rulesMatch = /## Rules(?<rulesText>.*?)## Deprecated Rules/ms.exec(readme);
```
Comment (fisker) [inline test/package.mjs:70]: Let's use the new mark to find the table

Comment (bmish) [reply test/package.mjs]: Updated to use the new comment markers.

Comment (fisker) [reply test/package.mjs]: We can also use `.split('|')` to replace the regex bellow.

**Comment by fisker** [comment]
@bmish  Would you like to send separate PR to generate `Usage` section too? 😍

File: test/package.mjs (hunk 12 above)
Comment (bmish) [reply test/package.mjs]: I just tried switching to that but it actually turned out to be less convenient with `.split('|')`. 
- We still need a regexp to separate the url and ID
- We need to trim the whitespace around the other parts
- We need to ignore the table header and blank lines

The regexp was handling most of that for us for free...

**Comment by bmish** [comment]
> @bmish Would you like to send separate PR to generate `Usage` section too? 😍

The `create-rule` script already adds new rules to that section so autogenerating it doesn't seem very beneficial to me...?

File: test/package.mjs (hunk 12 above)
Comment (fisker) [reply test/package.mjs]: Okay

**Review by fisker — APPROVED** [review]

**Comment by fisker** [comment]
`create-rule` always add new rule as `error`, and we also have external rule like `no-nested-ternary` overridden, also we'll turn rules like `unicorn/numeric-separators-style` to `"error"` soon, we'll need update two places. If we have this script, we only need update recommended config.

**Comment by fisker** [comment]
@sindresorhus Are we going to accept this?

**Comment by sindresorhus** [comment]
Yes

**Comment by fisker** [comment]
I think it's ready to merge.

**Comment by sindresorhus** [comment]
@bmish Is there any way we can force the first column to be wider?

**Comment by bmish** [comment]
> @bmish Is there any way we can force the first column to be wider?

I experimented and unfortunately didn't find a good way to do that.

**Comment by sindresorhus** [comment]
It seems like they managed to make it wider here somehow: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/README.md#supported-rules

**Comment by bmish** [comment]
> It seems like they managed to make it wider here somehow: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/README.md#supported-rules

Opened a fix in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1197.

**Comment by fisker** [comment]
Maybe we can use html instead of markdown table? `<td style="white-space: nowrap">`

**Comment by fisker** [comment]
Seems working

<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
      <th>✅</th>
      <th>🔧</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="white-space: nowrap">
        <a href="docs/rules/better-regex.md">better-regex long long long name</a>
      </td>
      <td>Improve regexes by making them shorter, consistent, and safer.</td>
      <td>✅</td>
      <td>🔧</td>
    </tr>
  </tbody>
</table>

```markdown
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
      <th>✅</th>
      <th>🔧</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="white-space: nowrap">
        <a href="docs/rules/better-regex.md">better-regex long long long name</a>
      </td>
      <td>Improve regexes by making them shorter, consistent, and safer.</td>
      <td>✅</td>
      <td>🔧</td>
    </tr>
  </tbody>
</table>
```

**Comment by bmish** [comment]
@fisker I just tested an HTML table in #1201 and it didn't help us because it turns out github markdown strips the style tag from `<td style="white-space: nowrap">`.

**Pull request merged**